### PR TITLE
Ports - Angela

### PR DESCRIPTION
**Pull request by AngelaOh**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | I struggled with figuring out what components of my code needed its own class or could simply be a method. For instance, I initially wanted hotel rooms and hotel blocks to be their own classes but decided against this after feeling accessing information would be more difficult. 
What was a design decision you made that changed over time over the project? | Over time, my understanding of my Reservation class changed to be more akin to Pending Reservations. This way, I was able to reuse that class for creating hotel blocks. 
What was a concept you gained clarity on, or a learning that you'd like to share? | I gained clarity on what my constructor method actually does. Every time I create an instance of a class, we can get creative with how certain instance variables instantiated. 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | Making two separate reservations on two different dates to expect the number of rooms available for one of those days to be one less than the usual 20. It is a nominal case because every time a reservation is made over a number of days, the number of available rooms for that range should decrease by one. 
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | Making a the check in and check out date the same but expecting the number of available rooms to still go down by one. This is an edge case because you would typically expect someone to stay at least one night. 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I feel that writing tests before writing the code felt much more organized and easier this time around. Pseudocoding first was still an issue as I was impatient to start writing code down as soon as I could.

### CODE REVIEW

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -1,8 +1,14 @@
...
+require "minitest/autorun"
+require "minitest/reporters"
+require "minitest/skip_dsl"
 
 Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new
 
 # Require_relative your lib files here!
+require_relative "../lib/reservation"
+require_relative "../lib/reservation_manager"
+require_relative "../lib/room"
```
Comment (kaidamasaki) [inline spec/spec_helper.rb]: It looks like you removed `lib/room.rb` but forgot to remove the `require_relative` here.  

This makes `rake` and `guard` unhappy.  In general it's a good idea to make sure to do a final test run _right before_ you turn something in.

File: lib/reservation.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
+require "date"
+require "pry"
+
+class Reservation
+  attr_accessor :reservation_id, :check_in, :check_out, :room, :is_hotel_blocker
+
+  def initialize(reservation_id, check_in: nil, check_out: nil, room: nil, is_hotel_blocker: false)
+    @reservation_id = reservation_id
+    check_in == nil ? @check_in = nil : @check_in = Date.parse(check_in)
```
Comment (kaidamasaki) [inline lib/reservation.rb]: It's easy for someone reading your code to overlook assignment inside of ternaries.

Consider:
```suggestion
    @check_in = check_in == nil ? nil : Date.parse(check_in)
```

File: lib/reservation.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    @room = all_rooms.sample
+    @is_hotel_blocker = is_hotel_blocker
+  end
+
+  def duration
+    if self.check_in != nil && self.check_out != nil
+      duration = (check_out - check_in).to_i
+
+      if duration <= 0
+        raise ArgumentError, "Check out time is not after check in time. Inputted check in date was #{check_in} and check out date was #{check_out}"
```
Comment (kaidamasaki) [inline lib/reservation.rb:21]: I like that you included the invalid dates in the message so users know what they did wrong!  😃

File: lib/reservation.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    check_out == nil ? @check_out = nil : @check_out = Date.parse(check_out)
+    @room = all_rooms.sample
+    @is_hotel_blocker = is_hotel_blocker
+  end
+
+  def duration
+    if self.check_in != nil && self.check_out != nil
+      duration = (check_out - check_in).to_i
+
+      if duration <= 0
```
Comment (kaidamasaki) [inline lib/reservation.rb]: Instead of calculating a duration as an `Integer` you can just compare the dates directly.

```suggestion
      if check_out <= check_in
```

File: lib/reservation_manager.rb [hunk 5]
```diff
@@ -0,0 +1,143 @@
+require "pry"
+require "date"
+
+class Reservation_Manager
+  attr_reader :rooms, :all_reservations, :available_rooms
+  attr_accessor :all_block_reservations
+
+  def initialize
+    @all_reservations = all_reservations || []
+    @all_block_reservations = all_block_reservations || []
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb:12]: Did you mean to pass in `all_reservations` and `all_block_reservations`?  

Lines 9 and 10 will just result in them both being empty arrays.  (Which seems like a totally reasonable way for them to start off.)

File: lib/reservation_manager.rb [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      end
+    else
+      raise ArgumentError, "Sorry, we don't have enough rooms available for your rooms in a block."
+    end
+
+    return block_rooms
+  end
+
+  def find_available_rooms(finding_check_in, finding_check_out)
+    available_rooms = (1..20).map { |i| i }
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb]: Not at all a big deal but this would be a little clearer as:

```suggestion
    available_rooms = (1..20).to_a
```

File: lib/reservation_manager.rb [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    date1 = Date.parse(finding_check_in)
+    date2 = Date.parse(finding_check_out)
+    if date1 == date2
+      given_date_range = (date1..date2).to_a
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb]: Since we know both dates are equal consider:

```suggestion
      given_date_range = [date1]
```

File: lib/reservation_manager.rb [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    else
+      reservation = Reservation.new(1)
+      reservation.check_in = fixed_check_in
+      reservation.check_out = fixed_check_out
+
+      available_rooms = find_available_rooms(check_in, check_out)
+      if available_rooms.length > 0
+        reservation.room = available_rooms[0]
+      else
+        raise ArgumentError, "There are no available rooms at the moment. Please try again later!"
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb]: An `ArgumentError` doesn't make sense here because the users's arguments are fine.  

You probably want a custom exception here instead to indicate there are no free rooms.

File: lib/reservation_manager.rb [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    available_block_rooms = find_available_rooms(check_in, check_out)
+    if available_block_rooms.length >= room_num
+      block_rooms = available_block_rooms.take(room_num)
+      block_rooms.each do |num|
+        new_block_room = Reservation.new(block_id, check_in: check_in, check_out: check_out, is_hotel_blocker: true)
+        new_block_room.room = num
+        all_block_reservations << new_block_room
+      end
+    else
+      raise ArgumentError, "Sorry, we don't have enough rooms available for your rooms in a block."
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb]: Same as above here, it's not the user's fault the hotel is full.

File: lib/reservation_manager.rb [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+          reservation.check_in = block_reservation.check_in
+          reservation.check_out = block_reservation.check_out
+        end
+      end
+
+      rooms_for_block = all_block_reservations.map { |block_instance| block_instance.room }
+      reservation.room = rooms_for_block[0]
+      all_block_reservations.shift
+    else
+      reservation = Reservation.new(1)
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb]: It looks like all of your reservations have an ID of `1`.

File: lib/reservation.rb [hunk 11]
```diff
@@ -0,0 +1,43 @@
+require "date"
+require "pry"
+
+class Reservation
+  attr_accessor :reservation_id, :check_in, :check_out, :room, :is_hotel_blocker
+
+  def initialize(reservation_id, check_in: nil, check_out: nil, room: nil, is_hotel_blocker: false)
```
Comment (kaidamasaki) [inline lib/reservation.rb]: I would like most of these to be mandatory instead of being set after initializing the instance.

```suggestion
  def initialize(reservation_id, check_in:, check_out:, room:, is_hotel_blocker: false)
```

Right now it's possible to create a reservation with no room or date range associated with it.

File: lib/reservation_manager.rb [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+      reservation.check_in = fixed_check_in
+      reservation.check_out = fixed_check_out
+
+      available_rooms = find_available_rooms(check_in, check_out)
+      if available_rooms.length > 0
+        reservation.room = available_rooms[0]
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb:39]: I would have liked to see this passed into `Reservation.new` instead of added after the fact.

File: lib/reservation_manager.rb [hunk 13]
```diff
@@ -0,0 +1,143 @@
+require "pry"
+require "date"
+
+class Reservation_Manager
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb:4]: Minor style thing I'm not sure we mentioned in class.  When a class is multiple words you just use capital letters and not underscores.

```suggestion
class ReservationManager
```

File: lib/reservation_manager.rb [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+      else
+        raise ArgumentError, "There are no available rooms at the moment. Please try again later!"
+      end
+    end
+
+    all_reservations << reservation
+    return reservation
+  end
+
+  def make_hotel_block(block_id, check_in, check_out, room_num)
```
Comment (kaidamasaki) [inline lib/reservation_manager.rb:49]: I'd like to see the `Reservation_Manager` handle `block_id`s for the user.  That way a user can't accidentally give out the same `block_id` twice.

File: lib/reservation.rb [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+      cost_per_room = 150
+    else
+      cost_per_room = 200
+    end
+
+    reservation_cost = cost_per_room * duration
+    return reservation_cost
+  end
+
+  def all_rooms
```
Comment (kaidamasaki) [inline lib/reservation.rb:40]: This should probably live in `Reservation_Manager` since it doesn't relate to any specific `Reservation`.

You also generate the list of all rooms a few times in `Reservation_Manager` so that could clean up the code there.

**Comment by kaidamasaki** [comment]
# Hotel
## What We're Looking For



Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | Yes.
Answer comprehension questions | Yes.
**Design** |
Each class is responsible for a single piece of the program | Mostly.  Reservation handles both regular and block reservations.  (Which somewhat complicates the logic there.)
Classes are loosely coupled | Yes.
**Wave 1** |
List rooms | Sort of.  Room listing is handled by `Reservation` | Yes.
List reservations for a given date | Yes.
Calculate reservation price | Yes.  (Though it looks like you need to pass in the result of `duration` when calculating the cost.)
Invalid date range produces an error | Yes.  (Though only when checking the duration of the reservation.  Creating a reservation with an invalid date range is allowed.)
Test coverage | Yes.
**Wave 2** |
View available rooms for a given date range | Yes.
Reserving a room that is not available produces an error | Yes.  A free room will be assigned if available but if none are available an error is `raise`d.
Test coverage | Yes.
**Wave 3** |
Create a block of rooms | Yes.
Check if a block has rooms | No.
Reserve a room from a block | Yes.
Test coverage |
**Fundamentals** |
Names variables, classes and modules appropriately | Yes.
Understanding of variable scope - local vs instance | Yes.
Can create complex logical structures utilizing variables | Yes.
Appropriately uses methods to break down tasks into smaller simpler tasks | Yes.
Understands the differences between class and instance methods | N/A (architecture didn't require class methods)
Appropriately uses iterators and `Enumerable` methods | Yes.
Appropriately writes and utilizes classes | Mostly.  I think it could really have helped to split some of your Reservation functionality into multiple classes.
Appropriately utilizes modules as a namespace | Didn't use a module.  (That's fine.)
**Wrap Up** |
There is a refactors.txt file | No.
The file provides a roadmap to future changes | N/A
**Additional Feedback** | Good job!  <br><br>It seemed like you were somewhat unclear about where to put functionality.  (Both what methods should be part of a class and where you should validate data.)<br><br>That said while your solution has some issues but this was an extremely challenging assignment, congratulations for finishing all three waves!
